### ISSUE TITLE
p2p mode: "auth failure", "ProposalBlock is invalid" & "duplicate CONN"

### ISSUE THREAD

**Issue by swapnil-pandey** [tendermint/tendermint#2720]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
0.25.0-0c9c3292

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
kvstore

**Environment**:
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04

**What happened**:
I have four systems with ip: .83 - .86
I created a testnet by: `tendermint testnet --populate-persistent-peers --starting-ip-address 192.168.0.83 --v 4` and then copied the config folders to the respective systems. I have set the creation of empty blocks to false.
Now when I run `tendermint node` on all four systems, tendermint is unable to commit the initial blocks and is stuck on height 2, forever increasing the round. On digging up the logs there are three types of errors popping up frequently: 

1. `E[10-29|13:50:14.971] enterPrevote: ProposalBlock is invalid       module=consensus height=2 round=93 err="Wrong Block.Header.AppHash.  Expected 8802000000000000, got 0000000000000000"`

2. `I[10-29|12:07:57.921] Error reconnecting to peer. Trying again     module=p2p tries=8 err="duplicate CONN<192.168.0.86:26656>: %!s(<nil>)" 
addr=0a97e401f8516b205d10ccc72abe0f59d15165de@192.168.0.86:26656`

3. `I[10-29|12:07:57.969] Inbound Peer rejected                        module=p2p err="auth failure: secrect conn failed: EOF" numPeers=2`

**What you expected to happen**:
I expected to get a functioning blockchain with 4 nodes.

**Comment by ebuchman** [comment]
The important error here is this one:

> E[10-29|13:50:14.971] enterPrevote: ProposalBlock is invalid module=consensus height=2 round=93 err="Wrong Block.Header.AppHash. Expected 8802000000000000, got 0000000000000000"

It suggests that the nodes did not all start from the same state. Can you confirm there was no stale data on any of the nodes and that they started from a clean slate with no `data` dir? It looks like one of them may have data from an old kvstore lying around.

**Comment by swapnil-pandey** [comment]
@ebuchman  before starting the nodes, I did delete everything present at `~/.tendermint` except the config folder on all the nodes.

**Comment by ebuchman** [comment]
It looks like you're running the `kvstore` in a separate process? How are you starting it? And would it be possible to attach complete logs (or at least the first block and the first few rounds of height 2) for one or some of the nodes?

**Comment by ebuchman** [comment]
For some reason after committing the first block that node computes a bad app hash:

> I[10-30|10:47:52.065] Committed state                              module=state height=1 txs=0 appHash=8802000000000000

Is it possible the `kvstore` was left running from a previous instance and not restarted, and so it has stale data?

**Comment by swapnil-pandey** [comment]
Should the `kvstore` abci-cli command be run every time afresh?

**Comment by ebuchman** [comment]
If you are restarting the blockchain from 0, yes, absolutely. Otherwise the kvstore will think you're at height X while Tendermint starts over from 0

**Comment by ebuchman** [comment]
Closing this as it seems its resolved. Please let us know, or open a new issue, if not.

**Comment by swapnil-pandey** [comment]
But what if we have an already running blockchain and a new node wants to join it? We cannot start the blockchain afresh every time a new peer joins right? How can we tackle that situation?

**Comment by ebuchman** [comment]
That should not be a problem - a new peer can join just fine. It should start from a fresh state using the same genesis block and sync right up. Other nodes do not need to restart anything in this case.

Only if you're starting a new blockchain do you need to reset the state

**Comment by njmurarka** [comment]
> The important error here is this one:
> 
> > E[10-29|13:50:14.971] enterPrevote: ProposalBlock is invalid module=consensus height=2 round=93 err="Wrong Block.Header.AppHash. Expected 8802000000000000, got 0000000000000000"
> 
> It suggests that the nodes did not all start from the same state. Can you confirm there was no stale data on any of the nodes and that they started from a clean slate with no `data` dir? It looks like one of them may have data from an old kvstore lying around.

Silly question, perhaps, and apologies if this is way out of date. How can a node bring down the entire chain, even if it had stale data on that node? It seems highly worrisome if a single node (or even a bunch of them either by accident or intentionally) could do this... Am I missing something? Let's say I have a 100 validator network and someone with super-low power comes in with an intentionally or accidentally stale "data" dir. Is it possible this node could bring forth this issue?

I have run into a very similar situation (not exactly the same error message) 48 hrs ago and after reading your message, I was pondering if this is what happened. I had "reset" my chain (running on COSMOS) and when I started it up again, at some point, it did this. Perhaps one of the validators on my chain did not clean out their data folder? This comes to mind.
